### PR TITLE
`cp` progress bar implementation

### PR DESCRIPTION
**Pull request by Xoffio**
# NOTE
Clean duplicate of #7825
Sorry about all the mess guys... I got confuse with GitHub and and ended up mankind that mess.
This the same code I just cleaned the commits.

# Description
Progress bar implementation for the `cp` command. Now if the flag `-p` or `--progress` is set, then the user will be able to see the progress of the file or files being copy

progressbar_cp01
progressbar_cp02

# User-Facing Changes
A new flag (`--progress` `-p`) was added to the `cp` command 
Examples:
```nu
cp -p test_file.txt test_folder_1\
cp -r -p test_folder\* test_folder_1\
cp -r -p -i test_folder\* test_folder_1\
```

## Notes 
- The progress bar uses `std::io::{Read, Write}` instead of `std::fs::copy` to get the progress. that means that when the progress bar is used the copy process might be a bit slower.
- Progress bar for symbolic links TBD: Since symbolic links are usually very light I think is not worth it to duplicate the function `copy_symlink` just to add a progress bar that will be so fast to the point is not needed, but.. for consistency purposes we might need to added it, In that case I would have to pass the variable `progress` by parameter (to not duplicate code unnecessary). If I do that i would have to pass the `progress` var to every function to respect `copy_impl: impl Fn(PathBuf, PathBuf, Span)`. Please let me know if this is not clear :p

### CODE REVIEW

**Comment by rgwood** [comment]
Thanks! Looks like tests are failing.

**Comment by Xoffio** [comment]
> Thanks! Looks like tests are failing.

Yeah for some reason I am I can't get the test working even in the main branch and I can't check the code.. looking into it

File: crates/nu-command/tests/commands/cp.rs [hunk 1]
```diff
@@ -7,16 +7,42 @@ use nu_test_support::nu;
 use nu_test_support::playground::Playground;
 use std::path::Path;
 
+fn get_file_hash<T: std::fmt::Display>(file: T) -> String {
+    nu!("open {} | to text | hash md5", file).out
+}
```
Comment (rgwood) [inline crates/nu-command/tests/commands/cp.rs:12]: FYI, this is a bit expensive. Not gonna say we shouldn't do it this way, but `nu!` launches a new instance of Nushell and that's a lot slower than a function call. Probably OK to leave as is for now though.

**Review by rgwood — COMMENTED** [review]
Thank you for the PR! We appreciate you taking the time to get this ready, seems like it was a lot of work. Some comments:

## Testing

I'm a little unsure about the way the tests are implemented. It would be nice to eliminate some of the duplication/redundancy. For example, instead of this:

```rust 
fn copy_test() {
    nu!("cp ...");
    assert!(...)

    // same thing again but for --progress
    nu!("cp -p ...");
    assert!(...)
}
```
What about something like this?
```rust 
fn copy_test(progress: bool) {
    let progress_flag: String = if progress {"-p"} else {""};
    nu!("cp {progress_flag} ...");
    assert!(...)
}
```

I think the latter would be easier to maintain.

## 1 progress bar per file

I noticed that `cp -p` shows 1 progress bar per file:

image

This is a bit overwhelming when copying large files with `-r`. The progress bars are created and finished much too quickly for the user to read. Would it make more sense to show a single progress bar for all files?

## `ctrl+c` not supported

I noticed that `ctrl+c` doesn't work to interrupt a file copy with `-p`. We should probably fix that.

**Comment by Xoffio** [comment]
@rgwood Really appreciate you taking the time!!

> Test - I think the latter would be easier to maintain.

I love it! 100% agreed! I will be working on those changes.

> 1 progress bar per file

Is it ok to have both? I think the way is working right now is useful for some people... For example, I used to work for a VFX company where they manage a lot of big files. When artist wanted to move those files and work and the same time (because of crazy deadlines) it is important to know the status of each file. I am pretty sure they are other industries or cases likes this.

Maybe we can have two flags and the default is the clean one (the one you suggested )? 

> ctrl+c not supported

Could you check with out the `-p` flag. Right know i have that issue with my Nushell and I have the latest version.

**Comment by rgwood** [comment]
> Maybe we can have two flags and the default is the clean one (the one you suggested )?

Let's do just 1 flag. We try to keep the number of flags on commands to the bare minimum.

>it is important to know the status of each file

Would it be enough to have the following information in the progress display?

1. The name of the file currently being copied
2. The total bytes successfully copied (across all files)

> Could you check with out the `-p` flag. 

`ctrl+c` doesn't work without `-p`, but it should. Worth a (separate) issue.

**Comment by Xoffio** [comment]
> We try to keep the number of flags on commands to the bare minimum.

Totally understand 

> Would it be enough to have the following information in the progress display?

I think this might do it

>  it is a pretty bad UX for large numbers of files.

Yeah let's go for the cleaner way. Do you mind accepting this PR first ( after I fix the tests ) and then I can make it cleaner? I want to avoid dealing with merging issues if someone works on the cp file. I had a bad experience last time xd but if that is a problem that's ok too, it's just that it would be easier 

> ctrlc before copying each file. Should be easy,

Cool! I can do that too in a different PR!

**Comment by rgwood** [comment]
> Do you mind accepting this PR first ( after I fix the tests ) and then I can make it cleaner?

Sure 👍

**Comment by Xoffio** [comment]
Working on this today!

**Comment by Xoffio** [comment]
> fn copy_test(progress: bool) {
    let progress_flag: String = if progress {"-p"} else {""};
    nu!("cp {progress_flag} ...");
    assert!(...)
}

@rgwood  Actually we can't pass parameter when doing test. I get this error:
`functions used as tests can not have any arguments`
Maybe there is a way I am not aware of? If there is not, what do you think about a loop that goes through the two cp ?

**Comment by sholderbach** [comment]
> @rgwood Actually we can't pass parameter when doing test. I get this error: `functions used as tests can not have any arguments` Maybe there is a way I am not aware of? If there is not, what do you think about a loop that goes through the two cp ?

I think in some of the nushell crates we use `rstest` which has macros to do fixtures/variant testing like `pytest`. This might be an option if you have a more complex test setup.

**Comment by rgwood** [comment]
@Xoffio try something like:

```rust

#[test]
fn copy_test() {
  copy_test_impl(true);
  copy_test_impl(false);
}

fn copy_test_impl(progress: bool) {
    let progress_flag: String = if progress {"-p"} else {""};
    nu!("cp {progress_flag} ...");
    assert!(...)
}
```

**Comment by Xoffio** [comment]
@rgwood The tests are done! :D If you can we can merge this one and  I am starting to work to make it one progress bar. This will be a bit more work but it would be easy for me to merge later if someone works in the `cp` file

**Review by rgwood — APPROVED** [review]
Thanks for all the work on the tests. This looks like a step in the right direction, we appreciate it. Will probably merge soon once the dust settles from the last release.

**Pull request merged**